src/param_env.md

```diff
@@ -0,0 +1,17 @@
+# Parameter Environment
+
+When working with associated constants or generic types it is often relevant to
+have more information about the `Self` or generic parameters. Trait bounds and
+similar information is encoded in the `ParamEnv`. Often this is not enough
+information to obtain things like the type's `Layout`, but you can do all kinds
+of other checks on it (e.g. whether a type implements `Copy`) or you can
+evaluate an associated constant whose value does not depend on anything from the
+parameter environment.
+
+Although you can obtain a valid `ParamEnv` for any item via
+`tcx.param_env(def_id)`, this `ParamEnv` can be too generic for your use case.
+Using the `ParamEnv` from the surrounding context can allow you to evaluate more
+things.
+
+Another great thing about `ParamEnv` is that you can use it to eliminate the
+generic parameters from a `Ty` by calling `param_env.and(ty)`.
```

Comment:
That eliminates irrelevant bounds from the `param_env` based on `ty` (or anything else with types in it, like `Substs` or `Instance`), without changing the `ty`. cc @nikomatsakis

Reply:
Yes, this phrasing was confusing to me. Probably it makes sense to explain `ParamEnvAnv` a bit more -- basically, saying that often we need to pair up the environment with the thing we are operating on. When doing that, we use `param_env.and(foo)`. If, however, `foo` doesn't refer to generic parameters, we'll swap in an empty environment since the where clauses shouldn't matter (I'm not sure if this behavior is worth mentioning though; it's kind of a detail and anyhow I expect it to change at some point).